src/js/actions/layers.js

```diff
@@ -790,9 +790,40 @@ define(function (require, exports) {
     selectAll.post = [_verifyLayerSelection];
 
     /**
+     * Remove the given layers from the JavaScript model. This is useful from
+     * event handlers that tell us about layers which have already been removed
+     * from the model. To delete layers from a Photoshop document, use the
+     * deleteSelected action.
+     *
+     * @param {Document} document
+     * @param {Layer|Immutable.Iterable.<Layer>} layers
+     * @return {Promise}
+     */
+    var removeLayers = function (document, layers) {
+        if (layers instanceof Layer) {
+            layers = Immutable.List.of(layers);
+        }
+
+        return _getSelectedLayerIndices(document)
+            .bind(this)
+            .then(function (selectedLayerIndices) {
+                var payload = {
+                    documentID: document.id,
+                    layerIDs: collection.pluck(layers, "id"),
+                    selectedIndices: selectedLayerIndices
+                };
+
+                this.dispatch(events.document.history.nonOptimistic.DELETE_LAYERS, payload);
```

Comment:
I believe this scenario (clicking canvas to create a text layer and then ESCing before typing) causes two extraneous history states in our store that do not have corresponding PS history.  These are created by the layer actions `addLayers()` and `removeLayers()`, both of which dispatch history-changing events.

You can test this by creating a rectangle, changing its color twice, doing the point-text-empty-ESC dance, and then undoing twice.  You should see the rectangle change color on canvas, but not updated in the panel.

It seems like a simple fix for the removeLayers case (conditionally dispatching a non-history version of the event).  If it is not that simple for the addLayers case, and we therefore might need to do some retrospective history manipulation, then LMK.